### ISSUE TITLE
Add forEachAsClue extension

### ISSUE THREAD

**Issue by imanushin** [kotest/kotest#939]
**TLDR** - add ability to write something like this:
```kotlin
forbiddenUsers.forEachAsClue { forbiddenUser ->
    content shouldNot haveSubstring(forbiddenUser.name)
}
```

**Details:**

Sometimes it is better to check the result with the each argument from the list. Example:

```kotlin
forbiddenUsers.forEach { forbiddenUser ->
      content shouldNot haveSubstring(forbiddenUser.name)
}
```

However in this case error contains ```forbiddenUser.name``` only. To add whole user I have to do something like this:

```kotlin
forbiddenUsers.forEach { forbiddenUser ->
    forbiddenUser.asClue {
        content shouldNot haveSubstring(forbiddenUser.name)
        // or content shouldNot haveSubstring(it.name)
    }      
}
```

However last example has two user variables, which is not needed here.

So **proposal**: add extension to have code like this:

```kotlin
forbiddenUsers.forEachAsClue { forbiddenUser ->
    content shouldNot haveSubstring(forbiddenUser.name)
}
```

In this case ```forbiddenUser``` is added to the all errors. Moreover (possible we have this already), this function can aggregate all errors, so final test failure description can have all problematic ```forbiddenUser```.

So, function can be like this:

```kotlin
    fun <TValue> Iterable<TValue>.forEachAsClue1(action: (TValue) -> Unit) = forEach { element ->
        element.asClue {
            action(it)
        }
    }


    fun <TValue> Iterable<TValue>.forEachAsClue2(action: (TValue) -> Unit) {
        val errors = mapNotNull { element ->
            try {
                element.asClue {
                    action(it)
                }

                null
            } catch (e: Throwable) {
                e
            }
        }

        return when {
            errors.isEmpty() -> Unit
            errors.size == 1 -> throw errors.single()
            else -> {
                val firstError = errors.first()

                errors.drop(1).forEach {
                    firstError.addSuppressed(it)
                }

                throw firstError
            }
        }
    }
```

**Comment by sksamuel** [comment]
Added in master for 4.0 release.
